### PR TITLE
Sequentialization of nested interrupts

### PR DESCRIPTION
**Pull request by lihaol**
This series of patches adds support for sequentialization of nested interrupts to goto-instrument.

Usage: goto-instrument --scheduling-function f input.gb output.gb

Note: source code requires
1. a scheduling function that invokes ISRs;
2. a char\* array named __CPROVER_isr_array to store function names of ISRs, e.g. 
const char\* __CPROVER_isr_array[] = {"isrA", "isrB"};

3. functions __CPROVER_enable_isr(i) and __CPROVER_disable_isr(i) to enable and disable interrupt i, respectively. 

See 66815da for code examples.

### CODE REVIEW

File: src/goto-instrument/interrupt.cpp [hunk 1]
```diff
@@ -2,12 +2,12 @@
 
 Module: Interrupt Instrumentation
 
-Author: Daniel Kroening
-
-Date: September 2011
+Author: Daniel Kroening, Lihao Liang
```
Comment (tautschnig) [inline src/goto-instrument/interrupt.cpp]: Don't drop the Date

File: src/goto-instrument/interrupt.cpp [hunk 2]
```diff
@@ -2,12 +2,12 @@
...
 
 \*******************************************************************/
 
+#include <iostream>
+
```
Comment (tautschnig) [inline src/goto-instrument/interrupt.cpp]: Don't include this unguarded

File: src/goto-instrument/interrupt.cpp [hunk 3]
```diff
@@ -17,12 +17,13 @@ Date: September 2011
...
 
 #include "interrupt.h"
-#include "rw_set.h"
+#include "interrupt_util.h"
 
 #ifdef LOCAL_MAY
 #include <analyses/local_may_alias.h>
 #endif
 
+#if 0
```
Comment (tautschnig) [inline src/goto-instrument/interrupt.cpp]: No need for keeping any of this code if it has been fully moved.

**Comment by lihaol** [comment]
I have done a rebase and added regression tests.

**Comment by lihaol** [comment]
As CBMC cannot figure out the recursion in the scheduling function f(), I have to add --unwind 2 for cbmc in chain.sh. Would it be okay for the rest of the test cases in regression/goto-instrument?

**Comment by tautschnig** [comment]
@lihaol Please don't change chain.sh, but I think you can actually use --unwind 2 with goto-instrument thanks to @danpoe. Thus it should be safe to add --unwind 2 to the options in the respective test.desc file. Could you please give that a try and commit that one?

**Comment by lihaol** [comment]
@tautschnig I have tried to add --unwind 2 for goto-instrument in test.desc but CBMC keeps unwinding the recursion without terminating.

**Comment by tautschnig** [comment]
Which recursion is it actually that you are referring to?

**Comment by lihaol** [comment]
f() -> isrA() -> f()

**Comment by tautschnig** [comment]
Hmm, I see. It seems that you'll have to stick with a separate directory then. Apologies for having provided poor directions!

**Comment by lihaol** [comment]
No problem at all! I have done some final cleanup. Many thanks again for the reviews!

**Comment by tautschnig** [comment]
I'd say this is ready to be merged.

**Comment by hannes-steffenhagen-diffblue** [comment]
@tautschnig @lihaol This looks interesting but seems very out of date now. Do you think this would be worth reviving? Otherwise if there's no interest in this we'll be closing this soon for inactivity.

**Comment by martin-cs** [comment]
If revived then https://github.com/diffblue/cbmc/pull/48 should probably also be considered.

**Comment by TGWDB** [comment]
Closing this as too long being untouched and no response/interest in the last couple of months. Please feel free to reopen if this is going to be picked up or updated.

**Pull request closed without merging**